src/endpoint/s3/s3_bucket_policy_utils.js

```diff
@@ -145,33 +145,48 @@ async function has_bucket_policy_permission(policy, account, method, arn_path, r
-    for (const statement of statements) {
-        let action_fit = false;
-        let principal_fit = false;
-        let resource_fit = false;
-        let condition_fit = true;
-        for (const action of _.flatten([statement.Action])) {
-            dbg.log1('bucket_policy: action fit?', action, method);
-            if ((action === '*') || (action === 's3:*') || (action === method)) {
-                action_fit = true;
-            }
+function _is_action_fit(method, statement_action, isNotStatement) {
+    let action_fit = false;
+    for (const action of _.flatten([statement_action])) {
+        dbg.log1('bucket_policy: ', isNotStatement ? 'NotAction' : 'Action', ' fit?', action, method);
+        if ((action === '*') || (action === 's3:*') || (action === method)) {
+            action_fit = true;
         }
-        const statement_principal = statement.Principal.AWS ? statement.Principal.AWS : statement.Principal;
-        for (const principal of _.flatten([statement_principal])) {
-            dbg.log1('bucket_policy: principal fit?', principal, account);
-            if ((principal.unwrap() === '*') || (principal.unwrap() === account)) {
-                principal_fit = true;
-            }
+    }
+    return isNotStatement ? !action_fit : action_fit;
+}
+
+function _is_principal_fit(account, statement_principal, isNotStatement) {
+    let principal_fit = false;
+    statement_principal = statement_principal.AWS ? statement_principal.AWS : statement_principal;
+    for (const principal of _.flatten([statement_principal])) {
+        dbg.log1('bucket_policy: ', isNotStatement ? 'NotPrincipal' : 'Principal', ' fit?', principal, account);
+        if ((principal.unwrap() === '*') || (principal.unwrap() === account)) {
+            principal_fit = true;
         }
-        for (const resource of _.flatten([statement.Resource])) {
-            const resource_regex = RegExp(`^${resource.replace(qm_regex, '.?').replace(ar_regex, '.*')}$`);
-            dbg.log1('bucket_policy: resource fit?', resource_regex, arn_path);
-            if (resource_regex.test(arn_path)) {
-                resource_fit = true;
-            }
+    }
+    return isNotStatement ? !principal_fit : principal_fit;
+}
+
+function _is_resource_fit(arn_path, statement_resource, isNotStatement) {
+    let resource_fit = false;
+    for (const resource of _.flatten([statement_resource])) {
+        const resource_regex = RegExp(`^${resource.replace(qm_regex, '.?').replace(ar_regex, '.*')}$`);
+        dbg.log1('bucket_policy: ', isNotStatement ? 'NotResource' : 'Resource', ' fit?', resource_regex, arn_path);
+        if (resource_regex.test(arn_path)) {
+            resource_fit = true;
         }
-        condition_fit = await _is_condition_fit(statement, req, method);
+    }
+    return isNotStatement ? !resource_fit : resource_fit;
+}
+
+async function _is_statements_fit(statements, account, method, arn_path, req) {
```

Comment:
Would you please explain this function?
If we have one statement that fits then all statements fit?